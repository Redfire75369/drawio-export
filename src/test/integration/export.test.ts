import {configureToMatchImageSnapshot} from "jest-image-snapshot";
import {JimpMime, Jimp} from "jimp";
import {readFile} from "node:fs/promises";
import {join, normalize} from "node:path";
import {fileURLToPath} from "node:url";
import {vi} from "vitest";
import Exporter, {Format} from "../../index";

const toMatchImageSnapshot = configureToMatchImageSnapshot({
	comparisonMethod: "ssim",
	failureThresholdType: "percent",
	failureThreshold: 1,
});
expect.extend({toMatchImageSnapshot});

describe("export", () => {
	const fixtures = normalize(join(fileURLToPath(import.meta.url), "../../fixtures"));

	let input: string;
	let exporter: Exporter;

	vi.setConfig({testTimeout: 30000});

	beforeAll(async () => {
		input = await readFile(join(fixtures, "flowchart.drawio"), "utf-8");
		exporter = await Exporter.launch();
	});
	afterAll(async () => await exporter.close());

	it("exports the first page of the flowchart fixture to JPEG", async () => {
		expect.assertions(1);

		const image = await exporter.exportImage(input, 0, Format.JPEG);
		const converter = await Jimp.read(image);
		const converted = await converter.getBuffer(JimpMime.png);

		expect(converted).toMatchImageSnapshot();
	});

	it("exports the first page of the flowchart fixture to PNG", async () => {
		expect.assertions(1);

		const image = await exporter.exportImage(input, 0, Format.PNG);
		expect(image).toMatchImageSnapshot();
	});

	it("exports the second page of the flowchart fixture to PNG", async () => {
		expect.assertions(1);

		const image = await exporter.exportImage(input, 1, Format.PNG);
		expect(image).toMatchImageSnapshot();
	});

	it("exports the first page of the flowchart fixture to SVG", async () => {
		expect.assertions(1);

		const svg = await exporter.exportSvg(input, 0);

		expect(svg).toMatchSnapshot();
	});
});
