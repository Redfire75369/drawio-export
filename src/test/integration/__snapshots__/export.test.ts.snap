// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`export > exports the first page of the flowchart fixture to SVG 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="286px" height="396px" viewBox="-0.5 -0.5 286 396"><defs/><g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-0"><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-1"><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-2"><g><path d="M 62 42 L 62 81.88" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="stroke"/><path d="M 62 90.88 L 57.5 81.88 L 66.5 81.88 Z" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="all"/></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-3"><g><rect x="2" y="2" width="120" height="40" rx="6" ry="6" fill="#fff2cc" stroke="#d6b656" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 118px; height: 1px; padding-top: 22px; margin-left: 3px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 12px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;">Lamp doesn't work</div></div></div></foreignObject></g></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-4"><g><path d="M 62 172 L 62 201.88" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="stroke"/><path d="M 62 210.88 L 57.5 201.88 L 66.5 201.88 Z" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 191px; margin-left: 82px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 11px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: nowrap;">Yes</div></div></div></foreignObject></g></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-5"><g><path d="M 112 132 L 151.88 132" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="stroke"/><path d="M 160.88 132 L 151.88 136.5 L 151.88 127.5 Z" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 122px; margin-left: 137px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 11px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: nowrap;">No</div></div></div></foreignObject></g></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-6"><g><path d="M 62 92 L 112 132 L 62 172 L 12 132 Z" fill="#fff2cc" stroke="#d6b656" stroke-miterlimit="10" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 90px; height: 1px; padding-top: 130px; margin-left: 17px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 12px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;">Lamp<br />plugged in?</div></div></div></foreignObject></g></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-7"><g><rect x="162" y="112" width="120" height="40" rx="6" ry="6" fill="#fff2cc" stroke="#d6b656" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 118px; height: 1px; padding-top: 132px; margin-left: 163px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 12px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;">Plug in lamp</div></div></div></foreignObject></g></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-8"><g><path d="M 62 292 L 62 341.88" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="stroke"/><path d="M 62 350.88 L 57.5 341.88 L 66.5 341.88 Z" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 333px; margin-left: 82px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 11px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: nowrap;">No</div></div></div></foreignObject></g></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-9"><g><path d="M 112 252 L 151.88 252" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="stroke"/><path d="M 160.88 252 L 151.88 256.5 L 151.88 247.5 Z" fill="none" stroke="#d6b656" stroke-miterlimit="10" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 242px; margin-left: 137px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 11px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: nowrap;">Yes</div></div></div></foreignObject></g></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-10"><g><path d="M 62 212 L 112 252 L 62 292 L 12 252 Z" fill="#fff2cc" stroke="#d6b656" stroke-miterlimit="10" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 90px; height: 1px; padding-top: 250px; margin-left: 17px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 12px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;">Bulb<br />burned out?</div></div></div></foreignObject></g></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-11"><g><rect x="2" y="352" width="120" height="40" rx="6" ry="6" fill="#fff2cc" stroke="#d6b656" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 118px; height: 1px; padding-top: 372px; margin-left: 3px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 12px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;">Repair Lamp</div></div></div></foreignObject></g></g></g><g data-cell-id="WIyWlLk6GJQsqaUBKTNV-12"><g><rect x="162" y="232" width="120" height="40" rx="6" ry="6" fill="#fff2cc" stroke="#d6b656" pointer-events="all"/></g><g><g transform="translate(-0.5 -0.5)"><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 118px; height: 1px; padding-top: 252px; margin-left: 163px;"><div style="box-sizing: border-box; font-size: 0px; text-align: center;" data-drawio-colors="color: rgb(0, 0, 0); "><div style="display: inline-block; font-size: 12px; font-family: &quot;Helvetica&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: all; white-space: normal; overflow-wrap: normal;">Replace Bulb</div></div></div></foreignObject></g></g></g></g></g></g></svg>"
`;
